pkg/v1/tkg/client/upgrade_cluster.go

```diff
@@ -1068,56 +1064,46 @@ func (c *TkgClient) createVsphereInfrastructureTemplateForUpgrade(regionalCluste
-				  "imageTag": "%s"
-				}
-			  }
-			}
-		  }
-		}
-	  }`
-	patchKubernetesVersion := fmt.Sprintf(patchString,
-		clusterUpgradeConfig.UpgradeComponentInfo.KubernetesVersion,
-		clusterUpgradeConfig.UpgradeComponentInfo.KCPInfrastructureTemplateName,
-		clusterUpgradeConfig.UpgradeComponentInfo.KCPInfrastructureTemplateNamespace,
-		clusterUpgradeConfig.UpgradeComponentInfo.ImageRepository,
-		clusterUpgradeConfig.UpgradeComponentInfo.CoreDNSImageRepository,
-		clusterUpgradeConfig.UpgradeComponentInfo.CoreDNSImageTag,
-		clusterUpgradeConfig.UpgradeComponentInfo.EtcdImageRepository,
-		clusterUpgradeConfig.UpgradeComponentInfo.EtcdImageTag)
-
-	log.Infof("Applying KubeadmControlPlane Patch: %s", patchKubernetesVersion)
-
-	// Using polling to retry on any failed patch attempt. Sometimes if user upgrade
-	// workload cluster right after management cluster upgrade there is a chance
-	// that all controller pods are not started on management cluster
-	// and in this case patch fails. Retrying again should fix this issue.
-	pollOptions := &clusterclient.PollOptions{Interval: upgradePatchInterval, Timeout: upgradePatchTimeout}
-	err := regionalClusterClient.PatchResource(&capikubeadmv1beta1.KubeadmControlPlane{}, clusterUpgradeConfig.KCPObjectName, clusterUpgradeConfig.KCPObjectNamespace, patchKubernetesVersion, types.MergePatchType, pollOptions)
+	if ! (clusterUpgradeConfig.ActualComponentInfo.KubernetesVersion == clusterUpgradeConfig.UpgradeComponentInfo.KubernetesVersion &&
+		clusterUpgradeConfig.ActualComponentInfo.KCPInfrastructureTemplateName == clusterUpgradeConfig.UpgradeComponentInfo.KCPInfrastructureTemplateName) {
+		log.Infof("Updating the KCP object with k8s version %s", clusterUpgradeConfig.UpgradeComponentInfo.KubernetesVersion)
+
+		currentKCP.Spec.Version = clusterUpgradeConfig.UpgradeComponentInfo.KubernetesVersion
+		currentKCP.Spec.MachineTemplate.InfrastructureRef.Name = clusterUpgradeConfig.UpgradeComponentInfo.KCPInfrastructureTemplateName
+		currentKCP.Spec.MachineTemplate.InfrastructureRef.Namespace = clusterUpgradeConfig.UpgradeComponentInfo.KCPInfrastructureTemplateNamespace
+		currentKCP.Spec.KubeadmConfigSpec.ClusterConfiguration.ImageRepository = clusterUpgradeConfig.UpgradeComponentInfo.ImageRepository
+		currentKCP.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageRepository = clusterUpgradeConfig.UpgradeComponentInfo.CoreDNSImageRepository
+		currentKCP.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag = clusterUpgradeConfig.UpgradeComponentInfo.CoreDNSImageTag
+		currentKCP.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.Local.ImageRepository = clusterUpgradeConfig.UpgradeComponentInfo.EtcdImageRepository
+		currentKCP.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.Local.ImageTag = clusterUpgradeConfig.UpgradeComponentInfo.EtcdImageTag
+	}
+
+	err = regionalClusterClient.UpdateResource(currentKCP, clusterUpgradeConfig.KCPObjectName, clusterUpgradeConfig.KCPObjectNamespace)
```

Comment:
Not sure how important it is, but wanted to note that as we changed the implementation from `PatchResource` to `UpdateResource` we are changing the implementation to a single update instead of retrying multiple times on failure.
Should we consider adding `pollOptions *PollOptions` to the `UpdateResource` function to allow multiple retries? cc @vuil

Reply:
That's a good point. I don't think we should change the API signature as part of this patch. If we can manually have a polling mechanism, that'll be helpful

Reply:
@anujc25 Based on our discussion, I've added new function that uses polling